### PR TITLE
Refactor CheckpointSaver

### PR DESCRIPTION
**Pull request by hanlint**
This is in a series of PRs refactoring the checkpointing code for readability and modularity without changing the public-facing API.

This PR cleans up the checkpoint saver code. Touching the `checkpoint.save_checkpoints` function will be deferred to another PR.

### CODE REVIEW

File: docs/source/trainer/checkpointing.rst
Comment (mvpatel2000) [inline docs/source/trainer/checkpointing.rst:79]:
```suggestion
    >>> latest_checkpoint = trainer.saved_checkpoints[-1]
```

File: composer/utils/checkpoint.py
Comment (mvpatel2000) [inline composer/utils/checkpoint.py:78]: should we add `.pt` suffix to checkpoints?

File: composer/utils/checkpoint.py
Comment (mvpatel2000) [inline composer/utils/checkpoint.py:71]: Is this a suffix of a filepath? If so, can we name accordingly?

File: composer/callbacks/checkpoint_saver.py
Comment (mvpatel2000) [inline composer/callbacks/checkpoint_saver.py:375]:
```suggestion
        saved_path = checkpoint.save_checkpoint(
            state=state,
            filename=filename,
            weights_only=self.weights_only,
        )

        if not saved_path:  # not all ranks save
            return
```
nit: `saved` sounds like bool

File: composer/callbacks/checkpoint_saver.py
Comment (mvpatel2000) [inline composer/callbacks/checkpoint_saver.py:379]: should this `.lstrip('/')` be preset in `symlink` in new version?

**Review by mvpatel2000 — COMMENTED** [review]
Did a first pass, mostly nits. I think this looks good, but I'll do a deeper pass once tests are resolved

**Comment by hanlint** [comment]
@mvpatel2000 test are passing now!

File: composer/callbacks/checkpoint_saver.py
Comment (eracah) [inline composer/callbacks/checkpoint_saver.py:377]: If it is None, then there is no latest checkpoint path?

File: composer/callbacks/checkpoint_saver.py
Comment (eracah) [inline composer/callbacks/checkpoint_saver.py:409]: nit:
```suggestion
        self.saved_checkpoints.append(filename)
```
modifies list in place instead of creating a new one.

File: composer/callbacks/checkpoint_saver.py
Comment (eracah) [inline composer/callbacks/checkpoint_saver.py:411]:
```suggestion
        if len(self.saved_checkpoints) > self.num_checkpoints_to_keep
```
A bit more readable if we do this no?

File: composer/trainer/trainer.py
Comment (eracah) [inline composer/trainer/trainer.py:2369]: is it me or is this function confusingly named given that there is already a `save_checkpoint` and a `_save_checkpoint`? It looks recursive at first glance. any way to change the naming here without messing up the interface?

Comment (eracah) [inline composer/trainer/trainer.py:2369]: is it me or is this function confusingly named given that there is another `save_checkpoint` and a `_save_checkpoint`. Any way to change the naming here without breaking the interface?

Comment (mvpatel2000) [reply composer/trainer/trainer.py:2369]: Big +1. This is confusing

Comment (mvpatel2000) [reply composer/trainer/trainer.py:2369]: Resolving as it looks like duplicate comment (github glitch?)

**Comment by mvpatel2000** [comment]
We should hold this off until logging refactor gets in. At this point I think we should basically hold any PR that touches loglevel stuff or we're going to keep getting merge conflicts there. Evan's been going through enough merge hell lol

File: composer/utils/checkpoint.py
Comment (mvpatel2000) [inline composer/utils/checkpoint.py:469]: what does this mean? refactor what part?

**Review by mvpatel2000 — COMMENTED** [review]
Lgtm modulo previous comments + merge conflicts that will arise once logger destination is merged in

**Comment by eracah** [comment]
> We should hold this off until logging refactor gets in. At this point I think we should basically hold any PR that touches loglevel stuff or we're going to keep getting merge conflicts there. Evan's been going through enough merge hell lol

Haha I appreciate you, @mvpatel2000 !

File: composer/utils/checkpoint.py
Comment (hanlint) [reply composer/utils/checkpoint.py:469]: Will refactor this entire function in a follow-on PR :)

File: composer/trainer/trainer.py
Comment (hanlint) [reply composer/trainer/trainer.py:2400]: This is here because we wanted an easy way to for the user to save their checkpoints directly from the trainer,e.g.
```
trainer.save_checkpoint('my_checkpoint.pt')
```
instead of:
```
from composer.utils.checkpoint import save_checkpoint
save_checkpoint(trainer.state, 'my_checkpoint.pt')
```
which is a lot of unnecessary boilerplate. The user also has to remember where that utility function is being imported from everytime.

File: composer/callbacks/checkpoint_saver.py
Comment (hanlint) [reply composer/callbacks/checkpoint_saver.py:377]: Yes, non-zero ranks do not save, unless we are using deepspeed

File: composer/utils/checkpoint.py
Comment (hanlint) [reply composer/utils/checkpoint.py:71]: It's not just a suffix, but also handles the formatting (hence the word `Partial`). Does that make sense, or need a different name?

File: composer/utils/checkpoint.py
Comment (hanlint) [reply composer/utils/checkpoint.py:78]: I dont think we should automaticaly add -- but i believe the default filenames are all .pt

File: composer/callbacks/checkpoint_saver.py
Comment (hanlint) [reply composer/callbacks/checkpoint_saver.py:379]: yes its there:
```
symlink_name = self.latest_artifact_name.format(state, is_deepspeed).lstrip('/') + '.symlink'
```

File: composer/callbacks/checkpoint_saver.py
Comment (mvpatel2000) [inline composer/callbacks/checkpoint_saver.py:304]: if I pass `save_interval` as a callable, I think this change sets it to `None` now. I believe we need `self.save_interval = save_interval` outside the if statement

File: composer/utils/checkpoint.py
Comment (mvpatel2000) [reply composer/utils/checkpoint.py:78]: Optional nit: I'd like a comment which explains filename might already have a suffix or maybe change variable name? (Not sure if its necessary, but its a bit confusing to have a suffix be `''` since it's not clear the filename has `.pt` (potentially) already added)

**Review by mvpatel2000 — APPROVED** [review]
Approving to unblock! LGTM besides one issue with `self.save_interval`?

**Review by eracah — APPROVED** [review]

**Pull request merged**
